### DISCUSSION TITLE
.NET/C# Version

### DISCUSSION THREAD

**Discussion by rodrigopscampos** [open-spaced-repetition/free-spaced-repetition-scheduler#15] [Ideas]
I missed a .NET/C# version. Is there any roadmap for that ? I could help with it

**Comment by L-M-Sherlock** [comment]
I don't know C#. Contribution is welcome!

**Reply by MayorNG** [reply]
大佬好，我希望能够贡献一个FSRS的C#实现，目前我已经根据[ts-fsrs](https://github.com/open-spaced-repetition/ts-fsrs)制作了一个简单翻译成C#的版本[csharp_fsrs](https://github.com/MayorNG/csharp_fsrs)(FSRS-4.5)。虽然已经可以使用，但是计算结果与ts版本的计算结果在较长时间尺度下还是会存在误差（可能计算next_difficulty时四舍五入的问题）。

我希望能进一步完善我的代码，所以我想知道我应该依据哪些标准进行测试，以确保我的版本有在正确运行。以及是否存在其他我需要遵守的标准，比如何时需要将difficulty四舍五入。

（另外，我在我的版本中修改了一些ReviewLog的属性名，不然原本的属性名很容易让我感到混乱）

**Reply by L-M-Sherlock** [reply]
https://github.com/open-spaced-repetition/fsrs-rs/blob/3d0dd3a6c7d4c7e0f9d3a23a124a46665773af34/src/model.rs#L345

这里有写一些测试

**Reply by ishiko732** [reply]
- https://github.com/MayorNG/csharp_fsrs/blob/95f67878f0a3608f4ad63e321232165ec0c80f6f/csharp-fsrs/Scheduler.cs#L86-L91
这段代码有点问题，在`State.Review`情况下，需要改变`Again`卡片的`lapses`,这`lapses`参数便于设计暂停卡片feature。
- 还有不清楚C#传入的对象在内部是否不可变？目前`ts-fsrs`是保证入参的Card数据是不被改变的。

**Reply by tukao89** [reply]
I want to contribute the C# version, what should I do?

**Reply by L-M-Sherlock** [reply]
You can implement it in your repo. If it is production-ready, I will move it into our organization.

**Comment by tukao89** [comment]
@rodrigopscampos 
@L-M-Sherlock 
A dotnet core c# version is here https://github.com/tukao89/dotnet-fsrs

**Reply by L-M-Sherlock** [reply]
Do you mind maintaining this repo in our organization?

**Comment by shargon** [comment]
Any c# library updated?

**Reply by tukao89** [reply]
https://github.com/tukao89/dotnet-fsrs

**Reply by shargon** [reply]
Two years ago..
